nmtwizard/prepoperator.py

```diff
@@ -62,42 +64,264 @@ def __call__(self):
             yield tu_batch
 
 
-class FileWriter(object):
+@six.add_metaclass(abc.ABCMeta)
+class Consumer(object):
+    """Base class for using preprocess results."""
+
+    def __init__(self, result_dir):
+        self._result_dir = result_dir
+
+    @abc.abstractmethod
+    def __call__(self, tu_batch):
+        raise NotImplementedError()
+
+    def open_files(self, f):
+        pass
+
+    def close_files(self):
+        for _, obj in self.__dict__.items():
+            if hasattr(obj, "close") and callable(obj.close) and \
+               hasattr(obj, "closed") and not obj.closed:
+                obj.close()
+
+    def finalize(self, config):
+        pass
+
+
+class SubwordLearner(Consumer):
+
+    def __init__(self, config, result_dir):
+
+        super(SubwordLearner, self).__init__(result_dir)
+
+        self._subword_learners = {}
+
+        opt_multi = config.get('tokenization', {}).get('multi', {}).get('build_subword')
+        opt_source = config.get('tokenization', {}).get('source', {}).get('build_subword')
+        opt_target = config.get('tokenization', {}).get('target', {}).get('build_subword')
+
+        if opt_multi:
+            self._subword_learners['multi'] = tokenizer.make_subword_learner(opt_multi, result_dir)
+        if opt_source:
+            self._subword_learners['source'] = tokenizer.make_subword_learner(opt_source, result_dir)
+        if opt_target:
+            self._subword_learners['target'] = tokenizer.make_subword_learner(opt_target, result_dir)
+
+
+    def __call__(self, tu_batch):
+        # Feed lines to subword learners.
+        # TODO V2 : feed tokenized lines ?
+        # TODO V2 : undo all placeholder annotation for subword processing
+        for tu in tu_batch :
+            if 'source' in self._subword_learners:
+                self._subword_learners['source']['learner'].ingest(tu.src_raw)
+            if 'target' in self._subword_learners:
+                self._subword_learners['target']['learner'].ingest(tu.tgt_raw)
+            if 'multi' in self._subword_learners:
+                self._subword_learners['multi']['learner'].ingest(tu.src_raw)
+                self._subword_learners['multi']['learner'].ingest(tu.tgt_raw)
+
+
+    def finalize(self, config):
+        # Learn subword models and write them to files.
+        for side, learner in self._subword_learners.items():
+            name =  config['tokenization'][side]['build_subword']['name'] \
+                    if 'name' in config['tokenization'][side]['build_subword'] \
+                    else 'model'
+
+            subword_type = self._subword_learners[side]['subword_type']
+            size = self._subword_learners[side]['size']
+
+            if side == 'multi' :
+                out_file = os.path.join(self._result_dir, "joint_" + subword_type + \
```

Comment:
When there is more than 2 string concatenations, I suggest to use string formatting instead (e.g. `joint_%s_%s" % (subword_type, name)`)